### PR TITLE
Add lwip httpd example

### PR DESCRIPTION
**Pull request by peterharperuk**
threadsafe and freertos

Fixes https://github.com/raspberrypi/pico-examples/issues/266

### CODE REVIEW

**Comment by lurch** [comment]
```
make[2]: fork: Resource temporarily unavailable
```

Looks like the CI server ran out of processes? ping @liamfraser

**Comment by peterharperuk** [comment]
The sys version of this fails an assert in freertos. Not immediately obvious why

#0  _exit (status=status@entry=1) at /home/peterh/source/pico/pico-sdk/src/rp2_common/pico_runtime/runtime.c:183
#1  0x1000591e in __assert_func (file=file@entry=0x1001bd1c "/home/peterh/source/pico/freertos/FreeRTOS-Kernel/queue.c", 
    line=line@entry=1537, func=func@entry=0x1001c090 <__func__.5762> "xQueueSemaphoreTake", 
    failedexpr=failedexpr@entry=0x1001beb8 "pxQueue->uxItemSize == 0")
    at /home/peterh/source/pico/pico-sdk/src/rp2_common/pico_runtime/runtime.c:228
#2  0x100154e2 in xQueueSemaphoreTake (xQueue=xQueue@entry=0x20003280 <ucHeap+7076>, xTicksToWait=<optimised out>, 
    xTicksToWait@entry=4294967295) at /home/peterh/source/pico/freertos/FreeRTOS-Kernel/queue.c:1542
#3  0x10015664 in xQueueTakeMutexRecursive (xMutex=0x20003280 <ucHeap+7076>, xTicksToWait=xTicksToWait@entry=4294967295)
    at /home/peterh/source/pico/freertos/FreeRTOS-Kernel/queue.c:743
#4  0x10010e02 in sys_mutex_lock (mutex=mutex@entry=0x20001374 <mem_mutex>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/contrib/ports/freertos/sys_arch.c:220
#5  0x1000891e in mem_trim (rmem=rmem@entry=0x20022628 <ram_heap+616>, new_size=<optimised out>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/core/mem.c:751
#6  0x1000936e in pbuf_realloc (p=<optimised out>, new_len=<optimised out>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/core/pbuf.c:444
#7  0x10013b36 in mdns_send_outpacket (msg=msg@entry=0x20003748 <ucHeap+8300>, netif=netif@entry=0x200220cc <cyw43_state+2264>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/apps/mdns/mdns_out.c:789
#8  0x100116ce in mdns_send_probe (netif=0x200220cc <cyw43_state+2264>, destination=destination@entry=0x1001b768 <v4group>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/apps/mdns/mdns.c:2273
#9  0x10012464 in mdns_probe_and_announce (arg=0x200220cc <cyw43_state+2264>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/apps/mdns/mdns.c:2316
#10 0x1000d182 in sys_check_timeouts () at /home/peterh/source/pico/pico-sdk/lib/lwip/src/core/timeouts.c:390
#11 0x10010b66 in tcpip_timeouts_mbox_fetch (mbox=mbox@entry=0x200016cc <tcpip_mbox>, msg=msg@entry=0x200037ac <ucHeap+8400>)
    at /home/peterh/source/pico/pico-sdk/lib/lwip/src/api/tcpip.c:109
#12 0x10010b92 in tcpip_thread (arg=<optimised out>) at /home/peterh/source/pico/pico-sdk/lib/lwip/src/api/tcpip.c:142
#13 0x10014644 in xPortPendSVHandler ()
    at /home/peterh/source/pico/freertos/FreeRTOS-Kernel/portable/ThirdParty/GCC/RP2040/port.c:380

pxQueue->uxItemSize is 66

**Comment by peterharperuk** [comment]
@lurch Yes. Unfortunately that c file is generated by a binary that's built from c code. So the tool has to be built first. It didn't strike me as trivial to build a native binary in the sdk? I could check a windows and linux version in and generate picow_fsdata.c from that I guess? But it seems less fragile to check the source in and require a manual step if the files are changed. Ideally there would be a python version of the tool.

**Comment by lurch** [comment]
> It didn't strike me as trivial to build a native binary in the sdk?

AFAIK the SDK build-process _does_ build `pioasm` and `elf2uf2` as native binaries? But I've no idea if the mechanism used within `pico-sdk` is also usable by code within `pico-examples`, so I'll shut up now :wink: 

> But it seems less fragile to check the source in and require a manual step if the files are changed.

Yup.

**Comment by peterharperuk** [comment]
> AFAIK the SDK build-process does build pioasm and elf2uf2 as native binaries

Yes - I can vaguely see what it's doing. I might give it a try.

File: pico_w/freertos/httpd/picow_freertos_httpd.c [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 void httpd_init(void);
 
+static absolute_time_t connect_time;
```
Comment (lurch) [inline pico_w/freertos/httpd/picow_freertos_httpd.c:20]: Looking at the code, this seems to be "wifi connection time"? I wonder if people might confuse it for "HTTP-request connection time"? :shrug:

**Comment by peterharperuk** [comment]
Currently failing...

*** PANIC ***                                                                                        
sys_timeout: timeout != NULL, pool MEMP_SYS_TIMEOUT is empty

**Comment by peterharperuk** [comment]
Needs https://github.com/georgerobotics/cyw43-driver/pull/39

**Comment by peterharperuk** [comment]
Now needs https://github.com/raspberrypi/pico-sdk/issues/1139 :)

**Comment by peterharperuk** [comment]
I think this is ready!

**Comment by peterharperuk** [comment]
Fails on Windows

ninja: error: 'pico_w/freertos/httpd/makefsdata/makefsdata.exe', needed by 'pico_w/freertos/httpd/content/picow_fsdata.c', missing and no known rule to make it

**Comment by peterharperuk** [comment]
ah ha - fails on a proper OS as well when using ninja

**Comment by peterharperuk** [comment]
ninja works on ubuntu now. Still can't get Windows to build a native binary. I think I'll give up on this.

**Pull request closed without merging**